packages/relay-runtime/network/__tests__/RelayObservable-test.js

```diff
@@ -1444,6 +1444,28 @@ describe('RelayObservable', () => {
       expect(list).toEqual(['error', error]);
     });
 
+    it('Converts an AsyncGenerator to an Observable', done => {
+      const value = {key: 'value'};
+
+      async function* generate() {
+        yield value;
+      }
+      const result = generate();
+      const obs = RelayObservable.from(result);
+
+      obs.subscribe({
+        next: val => {
+          expect(val).toEqual(value);
+        },
+        error: err => {
+          done(err);
+        },
+        complete: () => {
+          done();
+        },
+      });
```

Comment:
This approach to testing isn't very reliable, because it doesn't guarantee that next() is actually called. Our approach is to store all the events into an array and then assert the final result of the array, which checks that all the right events occurred and in the right order. Something like:

```javascript
it('Converts an AsyncGenerator to an Observable', () => {
  const value = ...;
  // ... same
  const obs = RelayObservable.from(result);
  
  // changed part
  const events = [];
  obs.subscribe({
    next: val => events.push(...['next', val]),
    error: err => events.push(...['error', err]),
    complete: () => events.push('complete'),
});
jest.runAllTimers();
expect(events).toEqual(['next', value, 'complete']);
```

Reply:
WOW, thank you so much for the guidance! Tricky stuff to test this on my pc. Nearly all snapshots fail:
![image](https://user-images.githubusercontent.com/599459/105418977-b57bda80-5c89-11eb-9261-70bc89da9f2e.png)

---

but ya, love the approach.

Reply:
Maybe you can help me with this, because they are promises the `runAllTimers` does nothing in this case—so `events` is always an empty array. As the subscribe runs after the assertion.

I have amended the test to now also expect an assertion.